packages/next/src/server/use-cache/use-cache-wrapper.ts

```diff
@@ -450,20 +450,28 @@ async function getNthCacheEntry(
 
 async function encodeFormData(formData: FormData): Promise<string> {
   let result = ''
-  for (let [key, value] of formData) {
-    // We don't need this key to be serializable but from a security perspective it should not be
-    // possible to generate a string that looks the same from a different structure. To ensure this
-    // we need a delimeter between fields but just using a delimeter is not enough since a string
-    // might contain that delimeter. We use the length of each field as the delimeter to avoid
+
+  // We sort the form data entries to ensure that the order in which promise
+  // arguments are resolved does not affect the determinism of the cache key.
```

Comment:
Personally i'd add a bit more about the mechanics here, i think this explanation wouldn't be clear to someone who hasn't dug into the exact output of `encodeReply`

```suggestion
  // `encodeReply` will encode promises as multiple fields.
  //  so if our cached function args are`[promise1, promise2]`, it will produce something like this:
  //    field_0: ["$@1", "$@2"]
  //     ...
  //    field_1: <resolved value of promise1>
  //    field_2: <resolved value of promise2>
  // However, the order in which the 'resolved value' fields depends on the order in which the promises resolved,
  // which can be non-deterministic, and thus could result in unnecessarily different cache keys.
  // To ensure a deterministic order, we sort the form fields by their key, meaning that in the example above, `field_1` would always come before `field_2`.
```

Reply:
also maybe a permalink to the relevant react code?
https://github.com/facebook/react/blob/a35aaf704cca9a5db16f5b197e3ac17eb960b72f/packages/react-client/src/ReactFlightReplyClient.js#L491-L511